src/Orleans.Transactions/DistributedTM/TransactionInfo.cs

```diff
@@ -201,9 +194,40 @@ public override string ToString()
                 $"{TransactionId} {TimeStamp:o}",
                 (IsReadOnly ? " RO" : ""),
                 (OriginalException != null ? " Aborting" : ""),
-                $" {{{string.Join(" ", Participants.Select(kvp => $"{kvp.Key.ToShortString()}:{kvp.Value.Reads},{kvp.Value.Writes}"))}}}",
-                TMCandidate != null ? $" TM={TMCandidate.ToShortString()}({TMBatchSize})" : ""
+                $" {{{string.Join(" ", this.Participants.Select(kvp => $"{kvp.Key}:{kvp.Value.Reads},{kvp.Value.Writes}"))}}}",
+                TMCandidate.Reference != null ? $" TM={TMCandidate.GetHashCode()}({TMBatchSize})" : ""
             );
         }
     }
+
+    public struct ParticipantId
+    {
+        public static IEqualityComparer<ParticipantId> Comparer = new IdComparer();
+
+        public GrainReference Reference;
```

Comment:
I think this struct is immutable, so we should express that in the definition.

With `Comparer`, we should also mark it `readonly`